VocaDbWeb/Scripts/Components/VdbPlayer/VdbPlayer.tsx

```diff
@@ -0,0 +1,459 @@
+			VdbPlayerConsole.debug(
+				`Playback ended (repeat mode: ${vdbPlayer.repeat})`,
+			);
+
+			const player = playerRef.current;
+
+			if (!player) return;
+
+			switch (vdbPlayer.repeat) {
+				case RepeatMode.One:
+					player.seekTo(0);
+					player.play();
+					break;
+
+				case RepeatMode.Off:
+				case RepeatMode.All:
+					if (playQueue.isLastItem) {
+						switch (vdbPlayer.repeat) {
+							case RepeatMode.Off:
+								vdbPlayer.setPlaying(false);
+								break;
+
+							case RepeatMode.All:
+								if (playQueue.hasMultipleItems) {
+									// HACK: Prevent vdbPlayer.next from being called in the same context.
+									// EmbedFile, EmbedNiconico, EmbedSoundCloud, EmbedYouTube and etc. must be rendered only after the `pv` prop has changed.
+									// Otherwise, the same PV may be played twice when switching between video services (e.g. Niconico => YouTube).
+									setTimeout(() => {
+										playQueue.goToFirst();
+									});
+								} else {
+									player.seekTo(0);
+									player.play();
+								}
+								break;
+						}
+					} else {
+						// HACK: Prevent vdbPlayer.next from being called in the same context.
+						// EmbedFile, EmbedNiconico, EmbedSoundCloud, EmbedYouTube and etc. must be rendered only after the `pv` prop has changed.
+						// Otherwise, the same PV may be played twice when switching between video services (e.g. Niconico => YouTube).
+						setTimeout(() => {
+							playQueue.next();
+						});
+					}
+					break;
+			}
+		}, [vdbPlayer, playQueue, playerRef]);
+
+		const options = React.useMemo(
+			() => ({
+				onError: handleError,
+				onPlay: handlePlay,
+				onPause: handlePause,
+				onEnded: handleEnded,
+			}),
+			[handleError, handlePlay, handlePause, handleEnded],
+		);
+
+		const handlePlayerChange = React.useCallback(
+			async (player?: PVPlayer) => {
+				try {
+					if (!player) return;
+
+					await player.load(pv.pvId);
+
+					player.play();
+				} catch (error) {
+					VdbPlayerConsole.error(
+						'Failed to load PV',
+						JSON.parse(JSON.stringify(pv)),
+						error,
+					);
+				}
+			},
+			[pv],
+		);
+
+		return (
+			<EmbedPV
+				pv={pv}
+				width="100%"
+				height="100%"
+				enableApi={true}
+				playerRef={playerRef}
+				options={options}
+				onPlayerChange={handlePlayerChange}
+			/>
+		);
+	},
+);
+
+const MiniPlayer = observer(
+	(): React.ReactElement => {
+		const { vdbPlayer, playQueue } = useVdbPlayer();
+
+		return (
+			<div
+				css={{
+					...(vdbPlayer.playerBounds === undefined
+						? {
+								position: 'fixed',
+								right: 0,
+								bottom: vdbPlayer.bottomBarVisible ? 50 : 0,
+								width: 16 * 25,
+								height: 9 * 25,
+								zIndex: 3939,
+						  }
+						: {
+								position: 'absolute',
+								left: vdbPlayer.playerBounds.x,
+								top: vdbPlayer.playerBounds.y,
+								width: vdbPlayer.playerBounds.width,
+								height: vdbPlayer.playerBounds.height,
+						  }),

```

Comment:
The "nostalgic diva" player should only be used in one place at any time. I really want to use [react-reverse-portal](https://github.com/httptoolkit/react-reverse-portal) to build a player once, and move it anywhere, but it didn't work well with [@vocadb/nostalgic-diva](https://github.com/VocaDB/nostalgic-diva) for some reason.